lib/galaxy/tools/execute.py

```diff
@@ -64,6 +64,8 @@ def execute_single_job(execution_slice, completed_job):
             message = EXECUTION_SUCCESS_MESSAGE % (tool.id, job.id, job_timer)
             log.debug(message)
             execution_tracker.record_success(execution_slice, job, result)
+            trans.sa_session.flush()
```

Comment:
Ahh... I didn't have this flush and a few workflow extraction tests were failing. Now I'm seeing that it was `` job.add_output_dataset_collection(output_name, collection_instance)`` below - that needs to be recorded before the expunge. A better solution is going to be just creating the output object directly like ImplicitCollectionJobsJobAssociation is done and skip the list update and flush I think. Step two - if we could just skip that object and the ImplicitCollectionJobsJobAssociation all together so they are not sitting around in the SA cache - maybe with executemany or something lower level - we could further reduce the memory consumption and speed things up I think.